File: go/0020.valid-parentheses/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:33
// leetgo: dev
// https://leetcode.com/problems/valid-parentheses/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func isValid(s string) bool {
	stack := []rune{}

	hash_map_open_to_close := map[rune]rune{
		'(': ')',
		'{': '}',
		'[': ']',
	}

	for _, val := range s {
		if val == '(' || val == '{' || val == '[' {
			stack = append(stack, val)
		} else {
			if len(stack) == 0 {
				return false
			}
			lastest_parenthesis := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			if hash_map_open_to_close[lastest_parenthesis] != val {
				return false
			}
		}
	}
	return len(stack) == 0
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	s := Deserialize[string](ReadLine(stdin))
	ans := isValid(s)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0020.valid-parentheses/solution.rs
// Created by Olgierd Palasz at 2026/08/04 20:45
// leetgo: dev
// https://leetcode.com/problems/valid-parentheses/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<u8> = Vec::with_capacity(s.len());

        for b in s.bytes() {
            match b {
                b'(' => stack.push(b')'),
                b'[' => stack.push(b']'),
                b'{' => stack.push(b'}'),
                _ => {
                    if stack.pop() != Some(b) {
                        return false;
                    }
                }
            }
        }

        stack.is_empty()
    }
}

// @lc code=end

fn main() -> Result<()> {
    let s: String = deserialize(&read_line()?)?;
    let ans: bool = Solution::is_valid(s);

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
